components/style/cgt_arena.rs

```diff
@@ -0,0 +1,361 @@
+            } else {
+                // The chunk should be allocated by someone else. Just make sure that
+                // has happened, spinning as necessary.
+                loop {
+                    if !self.chunks[chunk_idx].load(Ordering::Acquire).is_null() {
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    // Note: We could probably get away with Relaxed loads here because
+    // allocate() guarantees that the allocated data is observable on our thread,
+    // and it seems unlikely that the caller would be able to send the Token to
+    // another thread without some kind of SeqCst fence. But the benefit seems
+    // pretty marginal (since x86 is Acquire/Release by default), and the prospect
+    // of extremely rare ARM-only allocation races is pretty terrifying. So we just
+    // use the same atomic semantics at all callsites for the AtomicPtrs.
+    fn lookup(&self, idx: usize) -> &AtomicRefCell<Item> {
+        if idx >= ITEMS_PER_LIST {
+            // The item is in a subsequent list - traverse.
+            let list = self.next_list.load(Ordering::Acquire);
+            debug_assert!(!list.is_null());
+            return unsafe { (*list).lookup(idx - ITEMS_PER_LIST) };
+        }
+
+        // The item is in this list.
+        let chunk = idx / ITEMS_PER_CHUNK;
+        let chunk_idx = idx % ITEMS_PER_CHUNK;
+        let chunk_ptr = self.chunks[chunk].load(Ordering::Acquire);
+        debug_assert!(!chunk_ptr.is_null());
+        unsafe { &(*chunk_ptr).items[chunk_idx] }
+    }
+}
+
+/// `TokenData` is the internal data storage for an address within an `Arena`.
+/// It can be serialized to and from a `u64`, and used to construct a `Token`.
+#[derive(Clone, Copy, Debug)]
+pub struct TokenData {
+    generation: u32,
+    index: u32,
+}
+
+const LOW_MASK: u64 = 0xffffffff_u64;
+const HIGH_MASK: u64 = LOW_MASK << 32;
+
+impl TokenData {
+    /// Serializes the `TokenData` into a u64. The resulting value will always have
+    /// its low bit unset, which allows consumers to store one bit of additional
+    /// information in the serialization if they wish.
+    ///
+    /// If needed, this mechanism could be extended to allow consumers to allocate
+    /// an arbitrary number of user bits at the expense of reduced address space.
+    pub fn serialize(&self) -> u64 {
+        // Shift the generation into the high 32 bits.
+        let shifted_generation = (self.generation as u64) << 32;
+
+        // Shift the index left one bit to leave the bottom bit unused. This
+        // allows the callers to use tokens as tagged pointers.
+        debug_assert!(self.index & (1_u32 << 31) == 0);
+        let shifted_index = (self.index << 1) as u64;
+
+        shifted_generation | shifted_index
+    }
+
+    /// Deserializes a TokenData from a `u64`.
+    pub fn deserialize(x: u64) -> Self {
+        debug_assert!(x & 1 == 0, "Caller should strip the tag bit");
```

Comment:
It will just get shifted out. But anyway, none of this violates safety because Token::create is unsafe.

That said, it looks like there's a bug right now where deserialize doesn't unshift. I'll fix that and add a test.